capa/features/common.py

```diff
@@ -211,32 +212,29 @@ def evaluate(self, ctx, short_circuit=True):
                 raise ValueError("unexpected feature value type")
 
             if self.value in feature.value:
-                matches[feature.value].extend(locations)
+                matches[feature.value].update(locations)
                 if short_circuit:
                     # we found one matching string, thats sufficient to match.
                     # don't collect other matching strings in this mode.
                     break
 
         if matches:
-            # finalize: defaultdict -> dict
-            # which makes json serialization easier
-            matches = dict(matches)
-
             # collect all locations
             locations = set()
-            for s in matches.keys():
-                matches[s] = list(set(matches[s]))
-                locations.update(matches[s])
+            for locs in matches.values():
+                locations.update(locs)
 
             # unlike other features, we cannot return put a reference to `self` directly in a `Result`.
             # this is because `self` may match on many strings, so we can't stuff the matched value into it.
             # instead, return a new instance that has a reference to both the substring and the matched values.
-            return Result(True, _MatchedSubstring(self, matches), [], locations=locations)
+            return Result(True, _MatchedSubstring(self, dict(matches)), [], locations=locations)
         else:
             return Result(False, _MatchedSubstring(self, {}), [])
 
     def __str__(self):
-        return "substring(%s)" % self.value
+        v = self.value
+        assert isinstance(v, str)
```

Comment:
prove to mypy that `self.value` is a string. we'll see lots of additional checks added like this.

i use `assert` since it also adds a runtime guard that should never fire unless something is really wrong.

Reply:
good point, i'll check

Reply:
fixed in [aee0ec8](https://github.com/mandiant/capa/pull/1235/commits/aee0ec8016fd23850e3fe3ce777b84239ca6c616)